### DISCUSSION TITLE
Most efficient way to convert Scala case classes to js.Object?

### DISCUSSION THREAD

**Discussion by evbo** [circe/circe#2081] [Q&A] [answered]
Hi,

I was wondering if anyone can elaborate the docs a bit on `decodeJS` APIs in terms of how it might better enable using case classes in Scalaj.js code. For instance, there aren't many alternatives in Scala.js to easily use case classes when you need to convert to js.Object regularly. Some questions:
  - Why isn't there an `encodeJS` method for converting from case class to js.Object?
  - Is circe the best-in-class converter between js.Object and case class? Or is the overhead still a reason to just `extend js.Object`?

**Comment by armanbilge** [comment]
> Why isn't there an `encodeJS` method for converting from case class to js.Object?

The method you are looking for is here:

https://github.com/circe/circe/blob/79b3cd86fd06772c38b01ff6c2ff3eb7cd5e72cf/modules/scalajs/src/main/scala/io/circe/scalajs/package.scala#L60-L62

For any type `A` for which you have an `Encoder[A]`, you can convert it to a native JavaScript type. Furthermore, it is easy to obtain an `Encoder[A]` for a case class, using (semi-)automatic derivation.

---

> Is circe the best-in-class converter between js.Object and case class? Or is the overhead still a reason to just `extend js.Object`?

Probably not "best-in-class" :) this is not a usecase that Circe was directly designed for, but that happens to work.

In terms of overhead, it creates an intermediate `Json`:
```
MyCaseClass <=> circe.Json <=> js.Any
```
Depending on your needs, and the size of the data, this may or may not be acceptable.

**Comment by armanbilge** [comment, marked as answer]
> For instance, there aren't many alternatives in Scala.js to easily use case classes when you need to convert to js.Object regularly.

I recently became aware of this project:

- https://github.com/getshaka-org/native-converter

**Reply by evbo** [reply]
Both of your answers are A++, but I marked this one because that repo is EXACTLY the direction where I've hoped scalajs would end up!
